netaddr.go

```diff
@@ -510,25 +437,21 @@ func (ip IP) Prefix(bits uint8) (IPPrefix, error) {
 // Zone method to get it).
 // The ip zero value returns all zeroes.
 func (ip IP) As16() [16]byte {
-	if ip.ipImpl == nil {
-		return [16]byte{}
-	}
-	return ip.ipImpl.as16()
+	return ip.a
 }
 
 // As4 returns an IPv4 or IPv4-in-IPv6 address in its 4 byte representation.
 // If ip is the IP zero value or an IPv6 address, As4 panics.
 // Note that 0.0.0.0 is not the zero value.
 func (ip IP) As4() [4]byte {
-	if ip.ipImpl == nil {
+	switch ip.zt {
+	case z4:
+		return [4]byte{ip.a[o4+0], ip.a[o4+1], ip.a[o4+2], ip.a[o4+3]}
```

Comment:
Seems like we do this offset addition logic a fair amount, could be worth splitting out a method.